# **PURPOSE**
To analyze the dataset to deliver insights extracted

### **FIRST IMPORTING ENV**

In [203]:
!pip install numpy pandas matplotlib seaborn sklearn

In [204]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from statistics import *


 **LOADING THE DATA**

In [205]:
Path_to_data = "../data/telcom.csv"

In [206]:
#Read data from csv file
db = pd.read_csv(Path_to_data, na_values=['?', None])

#Display data
db.head()

Bearer Id            Start  Start ms              End  \
0 13,114,483,460,844,900,352.00   4/4/2019 12:01    770.00  4/25/2019 14:35   
1 13,114,483,482,878,900,224.00   4/9/2019 13:04    235.00   4/25/2019 8:15   
2 13,114,483,484,080,500,736.00   4/9/2019 17:42      1.00  4/25/2019 11:58   
3 13,114,483,485,442,799,616.00   4/10/2019 0:31    486.00   4/25/2019 7:36   
4 13,114,483,499,480,700,928.00  4/12/2019 20:10    565.00  4/25/2019 10:40   

   End ms    Dur. (ms)                   IMSI     MSISDN/Number  \
0  662.00 1,823,652.00 208,201,448,079,117.00 33,664,962,239.00   
1  606.00 1,365,104.00 208,201,909,211,140.00 33,681,854,413.00   
2  652.00 1,361,762.00 208,200,314,458,056.00 33,760,627,129.00   
3  171.00 1,321,509.00 208,201,402,342,131.00 33,750,343,200.00   
4  954.00 1,089,009.00 208,201,401,415,120.00 33,699,795,932.00   

                   IMEI     Last Location Name  ...  Youtube DL (Bytes)  \
0 35,521,209,507,511.00  9.16456699548519E+015  ...       15,854,611.00   
1 35,794,009,006,359.00                L77566A  ...       20,247,395.00   
2 35,281,510,359,387.00                D42335A  ...       19,725,661.00   
3 35,356,610,164,913.00                T21824A  ...       21,388,122.00   
4 35,407,009,745,539.00                D88865A  ...       15,259,380.00   

   Youtube UL (Bytes)  Netflix DL (Bytes)  Netflix UL (Bytes)  \
0        2,501,332.00        8,198,936.00        9,656,251.00   
1       19,111,729.00       18,338,413.00       17,227,132.00   
2       14,699,576.00       17,587,794.00        6,163,408.00   
3       15,146,643.00       13,994,646.00        1,097,942.00   
4       18,962,873.00       17,124,581.00          415,218.00   

   Gaming DL (Bytes)  Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0     278,082,303.00      14,344,150.00    171,744,450.00      8,814,393.00   
1     608,750,074.00       1,170,709.00    526,904,238.00     15,055,145.00   
2     229,584,621.00         395,630.00    410,692,588.00      4,215,763.00   
3     799,538,153.00      10,849,722.00    749,039,933.00     12,797,283.00   
4     527,707,248.00       3,529,801.00    550,709,500.00     13,910,322.00   

   Total UL (Bytes)  Total DL (Bytes)  
0     36,749,741.00    308,879,636.00  
1     53,800,391.00    653,384,965.00  
2     27,883,638.00    279,807,335.00  
3     43,324,218.00    846,028,530.00  
4     38,542,814.00    569,138,589.00  

[5 rows x 55 columns]

In [207]:
# column names
db.columns.tolist()

['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

EXPLORING THE DATAFRAME

In [208]:
# checking data types
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

In [209]:
# number of data points
print(f" There are {db.shape[0]} rows and {db.shape[1]} columns")

 There are 150001 rows and 55 columns


In [210]:
# Finding mode for manufacturer
db.mode()['Handset Manufacturer'][0]

'Apple'

In [211]:
# Finding mode for handset type
db.mode()['Handset Type'][0]

'Huawei B528S-23A'

In [212]:
# selecting columns of interest
manufacturer = db['Handset Manufacturer']
manufacturer.head()
manufacturer.describe()

count     149429
unique       170
top        Apple
freq       59565
Name: Handset Manufacturer, dtype: object

**Top 10 Handsets used by Customers**

In [213]:
# Getting top 10 handsets used by customers
n = 10
db['Handset Type'].value_counts()[:n].index.tolist()

['Huawei B528S-23A',
 'Apple iPhone 6S (A1688)',
 'Apple iPhone 6 (A1586)',
 'undefined',
 'Apple iPhone 7 (A1778)',
 'Apple iPhone Se (A1723)',
 'Apple iPhone 8 (A1905)',
 'Apple iPhone Xr (A2105)',
 'Samsung Galaxy S8 (Sm-G950F)',
 'Apple iPhone X (A1901)']

**Top 3 Handset Manufacturers**

In [214]:
# Getting the top 3 handset manufacturers
n = 3
manufacturer.value_counts()[:n].index.tolist()

['Apple', 'Samsung', 'Huawei']

**Most popular handset type**

In [215]:
mobitype = db['Handset Type']
mobitype.head()
mobitype.describe()

count               149429
unique                1396
top       Huawei B528S-23A
freq                 19752
Name: Handset Type, dtype: object

**APPLE BY ROW LOGIC**

In [216]:
# Using logic on rows to select only rows with apple as manufacturer
apple = db[manufacturer == 'Apple']
apple.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59565 entries, 22 to 149998
Data columns (total 55 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Bearer Id                                 59440 non-null  float64
 1   Start                                     59565 non-null  object 
 2   Start ms                                  59565 non-null  float64
 3   End                                       59565 non-null  object 
 4   End ms                                    59565 non-null  float64
 5   Dur. (ms)                                 59565 non-null  float64
 6   IMSI                                      59565 non-null  float64
 7   MSISDN/Number                             59464 non-null  float64
 8   IMEI                                      59565 non-null  float64
 9   Last Location Name                        59405 non-null  object 
 10  Avg RTT DL (ms)                 

**APPLE top 5**

In [217]:
# Getting the top 5 handset types in apple
n = 5
apple['Handset Type'].value_counts()[:n].index.tolist()

['Apple iPhone 6S (A1688)',
 'Apple iPhone 6 (A1586)',
 'Apple iPhone 7 (A1778)',
 'Apple iPhone Se (A1723)',
 'Apple iPhone 8 (A1905)']

**SAMSUNG BY ROW LOGIC**

In [218]:
# Using logic on rows to select only rows with samsung as manufacturer
samsung = db[manufacturer == 'Samsung']
samsung.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40839 entries, 0 to 149994
Data columns (total 55 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Bearer Id                                 40689 non-null  float64
 1   Start                                     40839 non-null  object 
 2   Start ms                                  40839 non-null  float64
 3   End                                       40839 non-null  object 
 4   End ms                                    40839 non-null  float64
 5   Dur. (ms)                                 40839 non-null  float64
 6   IMSI                                      40839 non-null  float64
 7   MSISDN/Number                             40579 non-null  float64
 8   IMEI                                      40839 non-null  float64
 9   Last Location Name                        40652 non-null  object 
 10  Avg RTT DL (ms)                  

**SAMSUNG top 5**

In [219]:
# Getting the top 5 handset types in samsung
n = 5
samsung['Handset Type'].value_counts()[:n].index.tolist()

['Samsung Galaxy S8 (Sm-G950F)',
 'Samsung Galaxy A5 Sm-A520F',
 'Samsung Galaxy J5 (Sm-J530)',
 'Samsung Galaxy J3 (Sm-J330)',
 'Samsung Galaxy S7 (Sm-G930X)']

**HUAWEI BY ROW LOGIC**

In [ ]:
# Using logic on rows to select only rows with huawei as manufacturer
huawei = db[manufacturer == 'Huawei']
huawei.info()

**HUAWEI top 5**

In [ ]:
# Getting the top 5 handset types in huawei
n = 5
huawei['Handset Type'].value_counts()[:n].index.tolist()

**INTEPRETATION AND RECOMMENDATION**

The top 3 phone manufactures by users are Apple(59565), Samsung(40839), and Huawei(34423). Marketing campaigns can therefore target these users and especially with the following handset types: 
Apple iPhone 6S (A1688), Apple iPhone 6 (A1586), Apple iPhone 7 (A1778), Apple iPhone Se (A1723), Apple iPhone 8 (A1905), Samsung Galaxy S8 (Sm-G950F), Samsung Galaxy A5 Sm-A520F, Samsung Galaxy J5 (Sm-J530), Samsung Galaxy J3 (Sm-J330), Samsung Galaxy S7 (Sm-G930X), Huawei B528S-23A, Huawei E5180, Huawei P20 Lite Huawei Nova 3E, Huawei P20, Huawei Y6 2018.

While the Apple brand is majorly popular and dorminantly used, Huawei B528S-23A is the most popular handset with 19752 customers.

# **ASSIGNMENT 1.1**

**Transforming the Data (Cleaning, Missing Values)**

In [ ]:
# summary statistics
db.describe()

In [ ]:
# checking data types
db.info()

In [ ]:
df_clean.columns.tolist()

In [ ]:
# how many missing values exist or better still what is the % of missing values in the dataset?
def percent_missing(df):

    # Calculate total number of cells in dataframe
    totalCells = np.product(df.shape)

    # Count number of missing values per column
    missingCount = df.isnull().sum()

    # Calculate total number of missing values
    totalMissing = missingCount.sum()

    # Calculate percentage of missing values
    print("The Telcom dataset contains", round(((totalMissing/totalCells) * 100), 2), "%", "missing values.")

percent_missing(db)

In [ ]:
# Checking for missing values
db.isna().sum()

In [ ]:
# drop columns with more than 30% missing values since filling this will greatly affect the dataset
df_clean = db.drop(['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)', 'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Nb of sec with 125000B < Vol DL', 'Nb of sec with 1250B < Vol UL < 6250B', 'Nb of sec with 31250B < Vol DL < 125000B', 'Nb of sec with 37500B < Vol UL', 'Nb of sec with 6250B < Vol DL < 31250B', 'Nb of sec with 6250B < Vol UL < 37500B',], axis=1)
df_clean.shape

In [ ]:
# dropping rows for columns with Unidue identifiers for customer data since we cannot fill missing values with existing data
df_clean.dropna(subset = ['MSISDN/Number'], inplace=True)
df_clean.dropna(subset = ['Bearer Id'], inplace=True)

**Filling Missing Values**

In [ ]:
# fill missing with ffill method for columns ('Nb of sec with Vol DL < 6250B','Nb of sec with Vol UL < 1250B', 'Avg RTT DL (ms)', 'Avg RTT UL (ms)', 'DL TP < 50 Kbps (%)', '50 Kbps < DL TP < 250 Kbps (%)', '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)', 'UL TP < 10 Kbps (%)', '10 Kbps < UL TP < 50 Kbps (%)', '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)')

def fix_missing_ffill(df, col):
    df[col] = df[col].fillna(method='ffill')
    return df[col]


def fix_missing_bfill(df, col):
    df[col] = df[col].fillna(method='bfill')
    return df[col]

df_clean['Nb of sec with Vol DL < 6250B'] = fix_missing_ffill(df_clean, 'Nb of sec with Vol DL < 6250B')
df_clean['Nb of sec with Vol UL < 1250B'] = fix_missing_ffill(df_clean, 'Nb of sec with Vol UL < 1250B')
df_clean['Avg RTT DL (ms)'] = fix_missing_ffill(df_clean, 'Avg RTT DL (ms)')
df_clean['Avg RTT UL (ms)'] = fix_missing_ffill(df_clean, 'Avg RTT UL (ms)')
#df_clean['DL TP < 50 Kbps (%)'] = fix_missing_ffill(df_clean, 'DL TP < 50 Kbps (%)')
#df_clean['50 Kbps < DL TP < 250 Kbps (%)'] = fix_missing_ffill(df_clean, '50 Kbps < DL TP < 250 Kbps (%)')
#df_clean['250 Kbps < DL TP < 1 Mbps (%)'] = fix_missing_ffill(df_clean, '250 Kbps < DL TP < 1 Mbps (%)')
#df_clean['DL TP > 1 Mbps (%)'] = fix_missing_ffill(df_clean, 'DL TP > 1 Mbps (%)')
#df_clean['UL TP < 10 Kbps (%)'] = fix_missing_ffill(df_clean, 'UL TP < 10 Kbps (%)')
#df_clean['10 Kbps < UL TP < 50 Kbps (%)'] = fix_missing_ffill(df_clean, '10 Kbps < UL TP < 50 Kbps (%)')
#df_clean['50 Kbps < UL TP < 300 Kbps (%)'] = fix_missing_ffill(df_clean, '50 Kbps < UL TP < 300 Kbps (%)')
#df_clean['UL TP > 300 Kbps (%)'] = fix_missing_ffill(df_clean, 'UL TP > 300 Kbps (%)')

# fill 'Handset Manufacturer' column with mode 
df_clean['Handset Manufacturer'] = df_clean['Handset Manufacturer'].fillna(df_clean['Handset Manufacturer'].mode()[0])
# fill 'Handset Type' column with mode 
df_clean['Handset Type'] = df_clean['Handset Type'].fillna(df_clean['Handset Type'].mode()[0])
# fill 'Last Location Name' column with mode 
df_clean['Last Location Name'] = df_clean['Last Location Name'].fillna(df_clean['Last Location Name'].mode()[0])

df_clean.describe()

In [ ]:
# Checking for missing values and then shape
df_clean.isna().sum()
df_clean.shape

In [ ]:
# Checking for missing values
df_clean.isna().sum()

**SELECT COLUMNS OF INTEREST AND GET AGGREGATE**


In [ ]:
#df_clean[['MSISDN/Number', 'Dur. (ms)', 'Social Media DL (Bytes)', 'Social Media UL (Bytes)', 'Google DL (Bytes)', 'Google UL (Bytes)', 'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 'Youtube UL (Bytes)', 'Netflix DL (Bytes)', Netflix UL (Bytes)', 'Gaming DL (Bytes)', 'Gaming UL (Bytes)', 'Other DL (Bytes)', 'Other UL (Bytes)', 'Total UL (Bytes)', 'Total DL (Bytes)']]

**No. of xDR Sessions**

In [ ]:
xDR = df_clean.groupby(['MSISDN/Number'])['Dur. (ms)'].count()
xDR.head()

**Session Durantion**

In [ ]:
# Sessions Durathion
Sessions = df_clean[['Dur. (ms)', 'Dur. (ms).1']]
Sessions.aggregate([np.sum, 'mean'], axis=0)
df_session_duration = df_clean.groupby(['IMSI'])['Dur. (ms)'].sum()
df_session_duration.head()

**Total UL and DL Data**

In [ ]:
# aggregate of the total download and upload data
Total_Data = df_clean[['Total UL (Bytes)', 'Total DL (Bytes)']]
Total_Data.aggregate([np.sum, 'mean'], axis=0)
df_total_data = df_clean.groupby(['IMSI'])['Total UL (Bytes)', 'Total DL (Bytes)'].sum()
df_total_data.head()

**Social Media Sessions**

In [ ]:
#aggregate of social media sessions
social_media = df_clean[['Social Media DL (Bytes)', 'Social Media UL (Bytes)']]
social_media.aggregate([np.sum, 'mean'], axis=0)
df_social = df_clean.groupby(['IMSI'])['Social Media DL (Bytes)', 'Social Media UL (Bytes)'].sum()
df_social.head()

**Google Sessions**

In [ ]:
#aggregate of google sessions
google = df_clean[['Google DL (Bytes)', 'Google UL (Bytes)']]
google.aggregate([np.sum, 'mean'], axis=0)
df_google = df_clean.groupby(['IMSI'])['Google DL (Bytes)', 'Google UL (Bytes)'].sum()
df_google.head()

**Email Sessions**

In [ ]:
#aggregate of email sessions
email = df_clean[['Email DL (Bytes)', 'Email UL (Bytes)']]
email.aggregate([np.sum, 'mean'], axis=0)
df_email = df_clean.groupby(['IMSI'])['Email DL (Bytes)', 'Email UL (Bytes)'].sum()
df_email.head()

**Youtube Sessions**

In [ ]:
#aggregate of youtube sessions
youtube = df_clean[['Youtube DL (Bytes)', 'Youtube UL (Bytes)']]
youtube.aggregate([np.sum, 'mean'], axis=0)
df_yt = df_clean.groupby(['IMSI'])['Youtube DL (Bytes)', 'Youtube UL (Bytes)'].sum()
df_yt.head()

**Netflix Sessions**

In [ ]:
#aggregate of netflix sessions
netflix = df_clean[['Netflix DL (Bytes)', 'Netflix UL (Bytes)']]
netflix.aggregate([np.sum, 'mean'], axis=0)
df_netflix = df_clean.groupby(['IMSI'])['Netflix DL (Bytes)', 'Netflix UL (Bytes)'].sum()
df_netflix.head()

**Gaming Session**

In [ ]:
#aggregate of gaming sessions
gaming = df_clean[['Gaming DL (Bytes)', 'Gaming UL (Bytes)']]
gaming.aggregate([np.sum, 'mean'], axis=0)
df_gaming = df_clean.groupby(['IMSI'])['Gaming DL (Bytes)', 'Gaming UL (Bytes)'].sum()
df_gaming.head()

**Other Sessions**

In [ ]:
#aggregate of other sessions
other = df_clean[['Other DL (Bytes)', 'Other UL (Bytes)']]
other.aggregate([np.sum, 'mean'], axis=0)
df_other = df_clean.groupby(['IMSI'])['Other DL (Bytes)', 'Other UL (Bytes)'].sum()
df_other.head()

In [ ]:
df_agg = df_clean[['IMSI', 'Dur. (ms)',
                'Social Media DL (Bytes)','Social Media UL (Bytes)', 'Google DL (Bytes)', 
                'Google UL (Bytes)', 'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 
                'Youtube UL (Bytes)', 'Netflix DL (Bytes)', 'Netflix UL (Bytes)', 'Gaming DL (Bytes)', 
                'Gaming UL (Bytes)', 'Other DL (Bytes)', 'Other UL (Bytes)', 'Total UL (Bytes)', 'Total DL (Bytes)']].groupby(['IMSI']).sum()

In [ ]:
df_agg.head()

# **ASSIGNMENT 1.2 - EDA**

**Replacing missing values with mean**

In [ ]:
df = df_agg.copy()
df.head()

In [ ]:
df.fillna(df.mean(), inplace = True)

In [ ]:
# Now which column(s) has missing values
df.isna().sum()

In [ ]:
percent_missing(df)

In [ ]:
df.describe()

**Taking care of Outliers** Fix later

In [ ]:
def format_float(value):
    return f'{value:,.2f}'

def convert_bytes_to_megabytes(df, bytes_data):
    megabyte = 1*10e+5
    df[bytes_data] = df[bytes_data] / megabyte
    return df[bytes_data]

def fix_outlier(df, column_list):
    for column in column_list:
        df[column] = np.where(df[column] > df[column].quantile(0.95), df[column].mean(),df[column])
    
    return df

pd.options.display.float_format = format_float

In [ ]:
df = fix_outlier(df, list(df.columns))

# **NON GRAPHICAL UNIVARIATE ANALYSIS**

Dispersion parameters, check and comment on variance, standard deviation and range. 

In [ ]:
df.columns.tolist()

In [ ]:
df.dtypes

In [ ]:
df.shape

**Analysis by Column**

In [ ]:
df.describe()

# **GRAPHICAL UNIVARIATE ANALYSIS **

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
###################################PLOTTING FUNCTIONS###################################

def plot_hist(df:pd.DataFrame, column:str, color:str)->None:
    # plt.figure(figsize=(15, 10))
    # fig, ax = plt.subplots(1, figsize=(12, 7))
    sns.displot(data=df, x=column, color=color, kde=True, height=7, aspect=2)
    plt.title(f'Distribution of {column}', size=20, fontweight='bold')
    plt.show()

def plot_count(df:pd.DataFrame, column:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.countplot(data=df, x=column)
    plt.title(f'Distribution of {column}', size=20, fontweight='bold')
    plt.show()
    
def plot_bar(df:pd.DataFrame, x_col:str, y_col:str, title:str, xlabel:str, ylabel:str)->None:
    plt.figure(figsize=(12, 7))
    sns.barplot(data = df, x=x_col, y=y_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.yticks( fontsize=14)
    plt.xlabel(xlabel, fontsize=16)
    plt.ylabel(ylabel, fontsize=16)
    plt.show()

def plot_heatmap(df:pd.DataFrame, title:str, cbar=False)->None:
    plt.figure(figsize=(12, 7))
    sns.heatmap(df, annot=True, cmap='viridis', vmin=0, vmax=1, fmt='.2f', linewidths=.7, cbar=cbar )
    plt.title(title, size=18, fontweight='bold')
    plt.show()

def plot_box(df:pd.DataFrame, x_col:str, title:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.boxplot(data = df, x=x_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.show()

def plot_box_multi(df:pd.DataFrame, x_col:str, y_col:str, title:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.boxplot(data = df, x=x_col, y=y_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.yticks( fontsize=14)
    plt.show()

def plot_scatter(df: pd.DataFrame, x_col: str, y_col: str, title: str, hue: str, style: str) -> None:
    plt.figure(figsize=(12, 7))
    sns.scatterplot(data = df, x=x_col, y=y_col, hue=hue, style=style)
    plt.title(title, size=20)
    plt.xticks(fontsize=14)
    plt.yticks( fontsize=14)
    plt.show()


In [ ]:
plot_box(df, "Dur. (ms)", "total user session duration outliers")

In [ ]:
plot_hist(df, "Total UL (Bytes)", "blue")

In [ ]:
plot_hist(df, "Total DL (Bytes)", "blue")

# **MULTIVARIATE ANALYSIS**

In [ ]:
def agg_data_per_app(df):
    cols = ['Social Media DL (Bytes)',
     'Social Media UL (Bytes)',
     'Google DL (Bytes)',
     'Google UL (Bytes)',
     'Email DL (Bytes)',
     'Email UL (Bytes)',
     'Youtube DL (Bytes)',
     'Youtube UL (Bytes)',
     'Netflix DL (Bytes)',
     'Netflix UL (Bytes)',
     'Gaming DL (Bytes)',
     'Gaming UL (Bytes)',
     'Other DL (Bytes)',
     'Other UL (Bytes)']
    df['Social Media'] = df[cols[0]] + df[cols[1]]
    df['Google'] = df[cols[2]] + df[cols[3]]
    df['Email'] = df[cols[4]] + df[cols[5]]
    df['Youtube'] = df[cols[6]] + df[cols[7]]
    df['Netflix'] = df[cols[8]] + df[cols[9]]
    df['Gaming'] = df[cols[10]] + df[cols[11]]
    df['Other'] = df[cols[12]] + df[cols[13]]
    return df
df = agg_data_per_app(df)
df.head()

In [ ]:
df_appdata = df[['Social Media','Google', 'Email','Youtube','Netflix','Gaming','Other']]
df_appdata.head()

In [ ]:
sns.pairplot(df_appdata.sample(5000), diag_kind = 'kde',
             plot_kws = {'alpha': 0.6, 's': 80, 'edgecolor': 'k'},
             height=4)